File: product-microservice/api/graphics_card_api.go
package api

import (
	"product/dto"
	"product/dto/filter"
	"product/mapper"
	"product/service"
	"strconv"

	"net/http"

	"github.com/gin-gonic/gin"
)


type GraphicsCardAPI struct {
	IGraphicsCardService service.IGraphicsCardService
}

func NewGraphicsCardAPI(graphicsCardService service.IGraphicsCardService) GraphicsCardAPI {
	return GraphicsCardAPI{IGraphicsCardService: graphicsCardService}
}

func (graphicsCardApi *GraphicsCardAPI) GetAll(c *gin.Context) {
	page, err := strconv.Atoi(c.Query("page"))
	pageSize, err := strconv.Atoi(c.Query("pageSize"))
    if err != nil {
		c.JSON(http.StatusBadRequest, err.Error())
		return
    }
	
	graphicsCards := graphicsCardApi.IGraphicsCardService.GetAll(page, pageSize)
	c.JSON(http.StatusOK, mapper.ToGraphicsCardDTOs(graphicsCards))
}

func (graphicsCardApi *GraphicsCardAPI) GetNumberOfRecords(c *gin.Context) {
	numberOfRecords := graphicsCardApi.IGraphicsCardService.GetNumberOfRecords()
	c.JSON(http.StatusOK, numberOfRecords)
}

func (graphicsCardApi *GraphicsCardAPI) GetByID(c *gin.Context) {
	id, err := strconv.Atoi(c.Param("id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, err.Error())
		return
	}
	
	graphicsCard, err := graphicsCardApi.IGraphicsCardService.GetById(id)
	
	if err == nil {
		c.JSON(http.StatusOK, mapper.ToGraphicsCardDTO(graphicsCard))
	} else {
		c.JSON(http.StatusBadRequest, err.Error())
	}
}

func (graphicsCardApi *GraphicsCardAPI) SearchByName(c *gin.Context) {
	page, err := strconv.Atoi(c.Query("page"))
	pageSize, err := strconv.Atoi(c.Query("pageSize"))
    if err != nil {
		c.JSON(http.StatusBadRequest, err.Error())
		return
    }

	graphicsCards, err := graphicsCardApi.IGraphicsCardService.SearchByName(page, pageSize, c.Query("name"))
	
	if err == nil {
		c.JSON(http.StatusOK, mapper.ToGraphicsCardDTOs(graphicsCards))
	} else {
		c.JSON(http.StatusBadRequest, err.Error())
	}
}

func (graphicsCardApi *GraphicsCardAPI) GetNumberOfRecordsSearch(c *gin.Context) {
	numberOfRecords := graphicsCardApi.IGraphicsCardService.GetNumberOfRecordsSearch(c.Query("name"))
	c.JSON(http.StatusOK, numberOfRecords)
}

func (graphicsCardApi *GraphicsCardAPI) Filter(c *gin.Context) {
	page, err := strconv.Atoi(c.Query("page"))
	pageSize, err := strconv.Atoi(c.Query("pageSize"))
    if err != nil {
		c.JSON(http.StatusBadRequest, err.Error())
		return
    }

	var filter filter.GraphicsCardFilter
	err = c.BindJSON(&filter)
	if err != nil {
		c.JSON(http.StatusBadRequest, err.Error())
		return
	}

	graphicsCards, err := graphicsCardApi.IGraphicsCardService.Filter(page, pageSize, filter)
	
	if err == nil {
		c.JSON(http.StatusOK, mapper.ToGraphicsCardDTOs(graphicsCards))
	} else {
		c.JSON(http.StatusBadRequest, err.Error())
	}
}

func (graphicsCardApi *GraphicsCardAPI) GetNumberOfRecordsFilter(c *gin.Context) {
	var filter filter.GraphicsCardFilter
	err := c.BindJSON(&filter)
	if err != nil {
		c.JSON(http.StatusBadRequest, err.Error())
		return
	}

	numberOfRecords := graphicsCardApi.IGraphicsCardService.GetNumberOfRecordsFilter(filter)
	c.JSON(http.StatusOK, numberOfRecords)
}

func (graphicsCardApi *GraphicsCardAPI) GetManufacturers(c *gin.Context) {
	manufacturers := graphicsCardApi.IGraphicsCardService.GetManufacturers()
	c.JSON(http.StatusOK, manufacturers)
}

func (graphicsCardApi *GraphicsCardAPI) GetChipManufacturers(c *gin.Context) {
	chipManufacturers := graphicsCardApi.IGraphicsCardService.GetChipManufacturers()
	c.JSON(http.StatusOK, chipManufacturers)
}

func (graphicsCardApi *GraphicsCardAPI) GetMemorySizes(c *gin.Context) {
	memorySizes := graphicsCardApi.IGraphicsCardService.GetMemorySizes()
	c.JSON(http.StatusOK, memorySizes)
}

func (graphicsCardApi *GraphicsCardAPI) GetMemoryTypes(c *gin.Context) {
	memoryTypes := graphicsCardApi.IGraphicsCardService.GetMemoryTypes()
	c.JSON(http.StatusOK, memoryTypes)
}

func (graphicsCardApi *GraphicsCardAPI) GetModelNames(c *gin.Context) {
	modelNames := graphicsCardApi.IGraphicsCardService.GetModelNames()
	c.JSON(http.StatusOK, modelNames)
}

func (graphicsCardApi *GraphicsCardAPI) Create(c *gin.Context) {
	var graphicsCardDTO dto.GraphicsCardDTO
	err := c.BindJSON(&graphicsCardDTO)
	if err != nil {
		c.JSON(http.StatusBadRequest, err.Error())
		return
	}

	graphicsCard := mapper.ToGraphicsCard(graphicsCardDTO)
	if err != nil {
		c.JSON(http.StatusBadRequest, err.Error())
		return
	}

	msg := graphicsCardApi.IGraphicsCardService.Create(graphicsCard)

	if msg == "" {
		c.JSON(http.StatusOK, "Graphics card added successfully.")
	} else {
		c.JSON(http.StatusBadRequest, msg)
	}
}

func (graphicsCardApi *GraphicsCardAPI) Update(c *gin.Context) {
	var graphicsCardDTO dto.GraphicsCardDTO
	err := c.BindJSON(&graphicsCardDTO)
	if err != nil {
		c.JSON(http.StatusBadRequest, err.Error())
		return
	}

	msg := graphicsCardApi.IGraphicsCardService.Update(graphicsCardDTO)

	if msg == "" {
		c.JSON(http.StatusOK, "Graphics card updated successfully.")
	} else  {
		c.JSON(http.StatusBadRequest, msg)
	} 
}

func (graphicsCardApi *GraphicsCardAPI) Delete(c *gin.Context) {
	id, err := strconv.Atoi(c.Param("id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, err.Error())
		return
	}
	
	error := graphicsCardApi.IGraphicsCardService.Delete(id)

	if error == nil {
		c.JSON(http.StatusOK, "Graphics card deleted successfully.")
	} else  {
		c.JSON(http.StatusBadRequest, error.Error())
	}
}

File: api-gateway/api/product_microservice/graphics_card_api.py
import requests
from flask import request, jsonify
from utils import token_utils, role
from utils.routes.product_microservice import graphics_card_api_routes


def get_all():
    args = request.args.to_dict()
    page = args.get("page")
    page_size = args.get("pageSize")
    try:
        r = requests.get(graphics_card_api_routes.BASE + graphics_card_api_routes.API +
                         graphics_card_api_routes.GET_ALL
                         + "?page={}&pageSize={}".format(page, page_size))
        resp = jsonify(r.json())
        resp.status_code = r.status_code
        return resp
    except requests.exceptions.RequestException as err:
        print(err)
        resp = jsonify(str(err))
        resp.status_code = 404
        return resp


def get_number_of_records():
    try:
        r = requests.get(graphics_card_api_routes.BASE + graphics_card_api_routes.API +
                         graphics_card_api_routes.GET_NUMBER_OF_RECORDS)
        resp = jsonify(r.json())
        resp.status_code = r.status_code
        return resp
    except requests.exceptions.RequestException as err:
        print(err)
        resp = jsonify(str(err))
        resp.status_code = 404
        return resp


def get_by_id(id):
    try:
        r = requests.get(graphics_card_api_routes.BASE + graphics_card_api_routes.API +
                         graphics_card_api_routes.GET_BY_ID + "/{}".format(id))
        resp = jsonify(r.json())
        resp.status_code = r.status_code
        return resp
    except requests.exceptions.RequestException as err:
        print(err)
        resp = jsonify(str(err))
        resp.status_code = 404
        return resp


def search_by_name():
    args = request.args.to_dict()
    page = args.get("page")
    page_size = args.get("pageSize")
    name = args.get("name")
    try:
        r = requests.get(graphics_card_api_routes.BASE + graphics_card_api_routes.API +
                         graphics_card_api_routes.SEARCH_BY_NAME
                         + "?page={}&pageSize={}&name={}".format(page, page_size, name))
        resp = jsonify(r.json())
        resp.status_code = r.status_code
        return resp
    except requests.exceptions.RequestException as err:
        print(err)
        resp = jsonify(str(err))
        resp.status_code = 404
        return resp


def get_number_of_records_search():
    args = request.args.to_dict()
    name = args.get("name")
    try:
        r = requests.get(graphics_card_api_routes.BASE + graphics_card_api_routes.API +
                         graphics_card_api_routes.GET_NUMBER_OF_RECORDS_SEARCH + "?name={}".format(name))
        resp = jsonify(r.json())
        resp.status_code = r.status_code
        return resp
    except requests.exceptions.RequestException as err:
        print(err)
        resp = jsonify(str(err))
        resp.status_code = 404
        return resp


def filter():
    args = request.args.to_dict()
    page = args.get("page")
    page_size = args.get("pageSize")
    data = request.json
    try:
        r = requests.post(graphics_card_api_routes.BASE + graphics_card_api_routes.API +
                          graphics_card_api_routes.FILTER +
                          "?page={}&pageSize={}".format(page, page_size), json=data)
        resp = jsonify(r.json())
        resp.status_code = r.status_code
        return resp
    except requests.exceptions.RequestException as err:
        print(err)
        resp = jsonify(str(err))
        resp.status_code = 404
        return resp


def get_number_of_records_filter():
    data = request.json
    try:
        r = requests.post(graphics_card_api_routes.BASE + graphics_card_api_routes.API +
                          graphics_card_api_routes.GET_NUMBER_OF_RECORDS_FILTER, json=data)
        resp = jsonify(r.json())
        resp.status_code = r.status_code
        return resp
    except requests.exceptions.RequestException as err:
        print(err)
        resp = jsonify(str(err))
        resp.status_code = 404
        return resp


def get_manufacturers():
    try:
        r = requests.get(graphics_card_api_routes.BASE + graphics_card_api_routes.API +
                         graphics_card_api_routes.GET_MANUFACTURERS)
        resp = jsonify(r.json())
        resp.status_code = r.status_code
        return resp
    except requests.exceptions.RequestException as err:
        print(err)
        resp = jsonify(str(err))
        resp.status_code = 404
        return resp


def get_chip_manufacturers():
    try:
        r = requests.get(graphics_card_api_routes.BASE + graphics_card_api_routes.API +
                         graphics_card_api_routes.GET_CHIP_MANUFACTURERS)
        resp = jsonify(r.json())
        resp.status_code = r.status_code
        return resp
    except requests.exceptions.RequestException as err:
        print(err)
        resp = jsonify(str(err))
        resp.status_code = 404
        return resp


def get_memory_sizes():
    try:
        r = requests.get(graphics_card_api_routes.BASE + graphics_card_api_routes.API +
                         graphics_card_api_routes.GET_MEMORY_SIZES)
        resp = jsonify(r.json())
        resp.status_code = r.status_code
        return resp
    except requests.exceptions.RequestException as err:
        print(err)
        resp = jsonify(str(err))
        resp.status_code = 404
        return resp


def get_memory_types():
    try:
        r = requests.get(graphics_card_api_routes.BASE + graphics_card_api_routes.API +
                         graphics_card_api_routes.GET_MEMORY_TYPES)
        resp = jsonify(r.json())
        resp.status_code = r.status_code
        return resp
    except requests.exceptions.RequestException as err:
        print(err)
        resp = jsonify(str(err))
        resp.status_code = 404
        return resp


def get_model_names():
    try:
        r = requests.get(graphics_card_api_routes.BASE + graphics_card_api_routes.API +
                         graphics_card_api_routes.GET_MODEL_NAMES)
        resp = jsonify(r.json())
        resp.status_code = r.status_code
        return resp
    except requests.exceptions.RequestException as err:
        print(err)
        resp = jsonify(str(err))
        resp.status_code = 404
        return resp


@token_utils.authorization_required(roles=[role.ROLE_EMPLOYEE])
def create():
    data = request.json
    headers = request.headers
    try:
        r = requests.post(graphics_card_api_routes.BASE + graphics_card_api_routes.API +
                          graphics_card_api_routes.CREATE, json=data, headers=headers)
        resp = jsonify(r.json())
        resp.status_code = r.status_code
        return resp
    except requests.exceptions.RequestException as err:
        print(err)
        resp = jsonify(str(err))
        resp.status_code = 404
        return resp


@token_utils.authorization_required(roles=[role.ROLE_EMPLOYEE])
def update():
    data = request.json
    headers = request.headers
    try:
        r = requests.put(graphics_card_api_routes.BASE + graphics_card_api_routes.API +
                         graphics_card_api_routes.UPDATE, json=data, headers=headers)
        resp = jsonify(r.json())
        resp.status_code = r.status_code
        return resp
    except requests.exceptions.RequestException as err:
        print(err)
        resp = jsonify(str(err))
        resp.status_code = 404
        return resp


@token_utils.authorization_required(roles=[role.ROLE_EMPLOYEE])
def delete(id):
    headers = request.headers
    try:
        r = requests.delete(graphics_card_api_routes.BASE + graphics_card_api_routes.API +
                            graphics_card_api_routes.DELETE + "/{}".format(id), headers=headers)
        resp = jsonify(r.json())
        resp.status_code = r.status_code
        return resp
    except requests.exceptions.RequestException as err:
        print(err)
        resp = jsonify(str(err))
        resp.status_code = 404
        return resp
